.cursor/rules/data-encoding.mdc

```diff
@@ -0,0 +1,292 @@
+                    "recipient" => recipient = Some(value),
+                }
+            }
+        }
+    }
+}
+```
+
+## Best Practices
+
+1. Input Validation:
+- Always validate decoded input before processing
+- Handle empty or malformed input gracefully
+- Provide clear error messages for decoding failures
+
+2. Error Handling:
+```rust
+// Use descriptive error messages
+.map_err(|e| format!("Failed to decode event log data: {}", e))?;
+
+// Handle specific error cases
+match decode_result {
+    Ok(data) => process_data(data),
+    Err(e) => handle_error(e),
+}
+```
+
+3. Type Safety:
+- Use generated Solidity types for contract interactions
+- Validate numeric conversions
+- Check array bounds and lengths
+- Verify address formats
+
+4. Output Formatting:
+```rust
+// For Ethereum submission
+let payload = TransactionPayload {
+    data: result.abi_encode(),
+    destination: validated_address,
+    value: amount,
+};
+
+// For CLI output
+let formatted_output = serde_json::to_string(&result)
+    .map_err(|e| format!("Failed to serialize output: {}", e))?;
+```
+
+## Data Size Limits
+
+1. Blockchain Constraints:
+```rust
+// Maximum transaction data size is ~24KB (24,576 bytes)
+// Recommended to stay under 5KB for optimal performance
+const MAX_RECOMMENDED_SIZE: usize = 5 * 1024; // 5KB
+
+// Gas costs per byte:
+// - 16 gas for non-zero bytes
+// - 4 gas for zero bytes
+```
+
+2. Best Practices for Large Datasets:
+```rust
+// Use pagination for API responses
+let url = format!(
+    "https://api.example.com/data?page={}&per_page=5",
+    page_number
+);
+
+// Limit response size
+if response.len() > MAX_RECOMMENDED_SIZE {
+    return Err("Response exceeds recommended size limit".to_string());
+}
+```
+
+3. Optimization Strategies:
+- Use pagination when fetching data from APIs
+- Limit number of items in response (e.g., top 5)
+- Remove optional fields if not essential
+- Consider data compression for large responses
+- Cache frequently requested data if possible
+- Use shorter field names in JSON responses
+
+
+## Common Encoding Scenarios
+
+1. Number Encoding:
+```rust
+// Input: Hex string to number
+let number = input.chars().next()
+    .and_then(|c| c.to_digit(16))
+    .map(|n| n as u64)
+    .ok_or("Invalid hex input")?;
+
+// Output: Number to bytes
+let bytes = number.to_string().into_bytes();
+```
+
+2. JSON Data:
+```rust
+// Decode JSON response
+let json: Root = fetch_json(req).await
+    .map_err(|e| e.to_string())?;
+
+// Encode structured data
+let output = serde_json::to_vec(&response_data)
+    .map_err(|e| e.to_string())?;
+```
+
+3. Contract Events:
+```rust
+// Define Solidity types
+sol! {
+    event NewTrigger(TriggerInfo _triggerInfo);
+    struct TriggerInfo {
+        uint64 triggerId;
+        bytes data;
+    }
+}
+
+// Encode for contract submission
+let encoded = solidity::DataWithId {
+    triggerId: id,
+    data: result.into()
+}.abi_encode();
```

Comment:
These do not match at all. If defined like this with the sol! macro, there is no `solidity` module.

Reply:
@evanorti For context on how this works

```
mod solidity {
    use alloy_sol_macro::sol;
    pub use ITypes::*;

    // The objects here will be generated automatically into Rust types.
    // If you update the .sol file, you must re-run `cargo build` to see the changes.
    // or restart your editor / language server.
    sol!("../../src/interfaces/ITypes.sol");
}
```

Since there is a module in the example it uses that `solidity::DataWithId` namespace

Reply:
If you are just using `sol!("../../src/interfaces/ITypes.sol");` without the mod, then you reference types like

```
use alloy_sol_macro::sol;
sol!("../../src/interfaces/ITypes.sol");

ITypes::NewTrigger
```